boa/docs/boa.8
.TH BOA 8 "Jan 22 2000" "Version 0.94"
.SH NAME
.B boa \- a single\-tasking high performance http server
.SH SYNOPSIS
.B boa
.RB [ -c 
.IR server_root ] 
.RB [ -r 
.IR chroot ] 
.RB [ -d ]
.SH DESCRIPTION
Boa is a single-tasking HTTP server. That means that unlike traditional web
servers, it does not fork for each incoming connection, nor does it fork many
copies of itself to handle multiple connections. It internally multiplexes all
of the ongoing HTTP connections, and forks only for CGI programs (which must be
separate processes.) Preliminary tests show Boa is more than twice as fast as
Apache.
.PP
The primary design goals of Boa are speed and security. Security, in the sense
of "can't be subverted by a malicious user", not "fine grained access control
and encrypted communications". Boa is not intended as a feature-packed server;
if you want one of those, check out WN from John Franks. Modifications to Boa
that improve its speed, security, robustness, and portability, are eagerly
sought. Other features may be added if they can be achieved without hurting the
primary goals. 
.SH OPTIONS
.IP \fB-d\fR 
instruct Boa not to fork itself (non-daemonize).

.IP "\fB-c \fIserver_root\fR"
choose a server root overriding the default SERVER_ROOT #define in 
.I defines.h

The server root must hold your local copy of the configuration file

.IP "\fB-r \fIchroot\fR"
instruct Boa where to chdir and chroot to.  The chdir/chroot
is done before the configuration file is read, or any log
files are opened.

.SH FILES
.TP
\fBboa.conf\fR \- the sole configuration file for Boa.  
The directives in this file are defined in the 
.B DIRECTIVES
section.

.TP
\fBmime.types\fR \- the 
MimeTypes <filename>
defines what Content-Type Boa will send in an HTTP/1.0
or better transaction.
.SH DIRECTIVES

The Boa configuration file is parsed with a lex/yacc or flex/bison generated
parser.  If it reports an error, the line number will be provided; it should
be easy to spot.  The syntax of each of these rules is very simple, and they
can occur in any order.  Where possible, these directives mimic those of NCSA
httpd 1.3; We saw no reason to introduce gratuitous differences.
.PP
Note: the "ServerRoot" is not in this configuration file.  It can be compiled
into the server (see 
.I defines.h
) or specified on the command line with the 
.B -c 
option.

The following directives are contained in the 
.I boa.conf
file, and most, but not all, are required.
.TP
Port <integer>
This is the port that Boa runs on.  The default port for http servers is 80. 
If it is less than 1024, the server must be started as root.
.TP
User <user name or UID>
The name or UID the server should run as.  For Boa to attempt this, the
server must be started as root.
.TP
Group <group name or GID>
The group name or GID the server should run as.  For Boa to attempt this,
the server must be started as root.
.TP
ServerAdmin <email address>
The email address where server problems should be sent.
Note: this is not currently used. 
.TP
ErrorLog <filename>
The location of the error log file.  If this does not start with
/, it is considered relative to the server root.
Set to /dev/null if you don't want errors logged. 
.TP
AccessLog <filename>
The location of the access log file.  If this does not start with /, it is
considered relative to the server root.
Comment out or set to /dev/null (less effective) to disable access logging.
.TP
VerboseCGILogs
This is a logical switch and does not take any parameters.
Comment out to disable.
.TP
ServerName <server_name>
The name of this server that should be sent back to
clients if different than that returned by gethostname.
.Tp
VirtualHost
This is a logical switch and does not take any parameters.
Comment out to disable.
Given DocumentRoot /var/www, requests on interface 'A' or IP 'IP-A'
become /var/www/IP-A.
Example: http://localhost/ becomes /var/www/127.0.0.1 
.TP
DocumentRoot <directory>
The root directory of the HTML documents. If this does not start with
/, it is considered relative to the server root.
.TP
UserDir <directory>
The name of the directory which is appended onto a user's home directory if a
~user request is received.
.TP
DirectoryIndex <filename>
Name of the file to use as a pre-written HTML directory index.  Please  make
and use these files.  On the fly creation of directory indexes can be slow.
.TP
DirectoryMaker <directory>
Name of the program used to generate on-the-fly directory listings.
The program must take one or two command-line arguments, the first
being the directory to index (absolute), and the second, which is optional,
contains what Boa would have the "title" of the document be.
Comment out if you don't want on the fly directory listings.
If this does not start with
/, it is considered relative to the server root.
.TP
KeepAliveMax <integer>
Number of KeepAlive requests to allow per connection.  Comment out, or set
to 0 to disable keepalive processing. 
.TP
KeepAliveTimeout <integer>
Number of seconds to wait before keepalive connections time out.
.TP
MimeTypes <file>
The location of the 
.I mime.types
file.  If this does not start with /, it is considered relative to 
the server root. Set to /dev/null if you do not want to load a mime types 
file. Do *not* comment out (better use AddType!)
.TP
DefaultType <mime type>
MIME type used if the file extension is unknown, or there is no file extension.
.TP
AddType <mime type> <extension> [extension...]
Associates a MIME type with an extension or extensions.
.TP
Redirect, Alias, and ScriptAlias <path1> <path2>
Redirect, Alias, and ScriptAlias all have the same semantics \-\- they
match the beginning of a request and take appropriate action.  Use
Redirect for other servers, Alias for the same server, and ScriptAlias to
enable directories for script execution.

Redirect allows you to tell clients about documents which used to exist
in your server's namespace, but do not anymore.  This allows you tell
the clients where to look for the relocated document.

Alias aliases one path to another.  Of course, symbolic links in the
file system work fine too.          

ScriptAlias maps a virtual path to a directory for serving scripts.
.PP
Please see the included 
.I boa.conf 
for defaults and examples.
.SH HISTORY
Like the Linux kernel, even numbered versions are "stable", and odd numbered
versions are "unstable", or rather, "development".
Versions 0.91 and 0.91beta of Boa were released by Paul Phillips <paulp@go2net.com>
.PP
Version 0.92 was released by Larry Doolittle on 
December 12, 1996.
.PP
Version 0.93 was the development version of 0.94. 
.PP
Version 0.94 was released 22 Jan 2000.
.SH BUGS
There are probably bugs, but we are not aware of any at this time.
.SH AUTHOR
Boa was created by Paul Phillips <paulp@go2net.com>. It is now being maintained and
enhanced by Larry Doolittle
<ldoolitt@boa.org> and
Jon Nelson <jnelson@boa.org>.
.PP
Linux is the development platform at the moment, other
OS's are known to work. If you'd like to
contribute to this effort, contact Larry or Jon via e-mail. 
.SH LICENSE
This program is distributed under the GNU General Public License, as noted in
each source file.


boa/ChangeLog
** Changes from 0.94.12 to 0.94.13
 * Change many instances of log_error_mesg + exit to DIE macro
 * Change all instance of log_error_mesg (without exit) to WARN macro
 * do a much better job of checking return values from malloc and
   especially strdup.
 * check results of calling umask and getrlimit
 * server_s is no longer a global int
 * check results of fork via switch instead of if (fork())
 * check for getopt.h and include it if found
 * remove unused #defines, and add WARN macro, and replace
   many calls to log_error_mesg(..) with WARN macro
 * fix bug in get_commonlog_time where time_offset calculation was
   the opposite of what it should be ('-' and '+' were swapped)
 * fix compatability bug with old and newer versions of flex/yacc
 * add check for AC_FUNC_MMAP to configure.in
 * fix really lame thinko in normalize_path, which would prepend the 
   results of earlier calls to results from later calls
 * Add MaxConnections, a configuration directive which allows the
   user to specify the maximum number of connections that Boa will
   accept concurrently.
 * add SERVER_ADDR and REQUEST_URI to environment of CGI
 * handle SIGBUS during writes of data that has been memory mapped
 * minor optimization in select.c that prevents DEAD requests from 
   being added to the block set
 * fix bug in CGI environment script_name - closes sf.net bug #576725
 * make 'status' variable local to requests.c, not local to every file
   by forgetting to declare 'extern' in globals.h :-|
 * make getsockname non-fatal, and do it every time because we may 
   need it for the CGI
 * some minor refactoring optimizations in hash.c
 
** Changes from 0.94.11 to 0.94.12
 * Renamed Changelog ChangeLog, and moved up to top-level directory
 * Next 3 items due in part or whole thanks to 
   Liam Widdowson
 * when printf'ing a pid type, force to int, because it could be 
   something else on other platforms.  Should probably change it to 
   a long, and use that.
 * backported chroot commandline support from 0.95
 * backported support for strdup, strstr, alphasort, and scandir
   from 0.95 
 * Fixed src/Makefile.in -- it didn't remove index_dir.o
 * backport create_temporary_file from 0.95
   (instead of using tmpnam)
 * Allow non-standard date format 31 September 2000 23:59:59 GMT
   Patch by Landon Curt Noll
 * Skip whitespace before HTTP/major.minor
   Adapted patch from Landon Curt Noll
 * open /dev/null first thing (affects chrooting)
 * properly handle sigalrm -- use sigalrm_flag and sigalrm_run
   instead of handling the signal in the signal handler
 * update manpage slightly
 * send 400 BAD Request when resource does not start with '/'
 * add grp.h to boa.h's includes -- remove from boa.c and config.c
 * removed duplicate header includes from boa.c, config.c, get.c, 
   ip.c, request.c, response.c
 * factor out creating the server socket and dropping privs
   into create_server_socket and drop_privs
 * type all functions in boa.c (except main) as static
 * set umask after opening /dev/null
 * tie stdin/stdout to /dev/null before commandline parse
 * removed old, unused chroot code
 * move builds_needs_escape earlier in the startup
 * move fork later in the startup
 * type all c_set_* as static in config.c
 * don't bother trying to change uid/gid (or error if the
   requested uid/gid doesn't exist) if not UID 0
 * return more appropriate error code when foo.html gives
   access denied, but foo.html.gz gives some other error
   (essentially report error associated with foo.html, not foo.html.gz)
 * send NOT Implemented when an unknown method is attempted
 * always attempt a 32k read right before close
   (stopgap until blackhole can be merged)
 * allow more than 1 space in logline between method, resource, 
   and http version
 * don't use inline functions
 * update configure.in so that autoconf 2.50 doesn't complain (as much)
 * properly use VPATH and srcdir according to autoconf docs
 * change curly-braces to parentheses in Makefilein
 * use $^ instead of manually listing the dependencies in Makefile.in
 * remove tests section in Makefile.in
 * write tags not TAGS in Makefile.in
 * Add gethostbyname and inet_aton to function checks
 * Add code from 0.95 which checks for socket in -lsocket, 
   inet_aton in -lresolv, and gethost{by}name in -lnsl
 * Also remove broken bc-based "how big is an unsigned int" checks:
   assume minimum of 32 bits and check in escape.c at runtime.
 * Added new file: README.chroot.solaris, based on a
   modified version by Liam Widdowson
 * Add check_struct_for.m4, which allows us to check a structure 
   for a member (found at http://www.gnu.org/software/ac-archive/
    authored by Wes Hardaker
 * Call "aclocal -I ." to rebuild aclocal.m4
 * Using new check-struct-for-member autoconf macro, check
   for tm_gmtoff and tm_zone in struct tm -- useful in
   portability tests for localtime.
 * Also check sockaddr_in for structure sin_len so we can set
   it properly.
 * index_dir.c (which ends up in boa_indexer) can now be compiled
   with USE_LOCALTIME, and if so, it will report the local time 
   using the timezone name. Otherwise it uses UTC time and UTC
   timezone designation.
 * fix buglet in mmap_cache.c which shows up when under
   heavy load by many different files.
   Found and squashed by Michal Kara
 * normalize paths on Aliases, log files, server root, dirmaker
   This makes sure that paths are 'absolute'
 * don't generate DOCUMENT_ROOT or SERVER_ROOT,
   CGIs have no business knowing that information
 * if CGI, chdir to the cgi's root path
   Bug found by Matt Callaway
 * remove ChrootPath and PidFile directives from the parser
   (they aren't used anyway)
 * keep track of maximum file descriptor in use to optimize call
   to select()
 * apply IPv6 patch from Jari Korva
 * optimize keep-alive copy data routine
 * try to use memcpy instead of strcpy/strcat in more places (alias.c)
 * update .depend file
 * use fcntl + GET_FL to get a file descriptor's flags, then

boa/src/boa.c
/*
 *  Boa, an http server
 *  Copyright (C) 1995 Paul Phillips <paulp@go2net.com>
 *  Some changes Copyright (C) 1996 Charles F. Randall <crandall@goldsys.com>
 *  Some changes Copyright (C) 1996 Larry Doolittle <ldoolitt@boa.org>
 *  Some changes Copyright (C) 1996-2002 Jon Nelson <jnelson@boa.org>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 1, or (at your option)
 *  any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

/* $Id: boa.c,v 1.99.2.2 2002/07/23 15:50:29 jnelson Exp $*/

#include "boa.h"
#include <sys/resource.h>

/* globals */
int backlog = SO_MAXCONN;
time_t start_time;

int sighup_flag = 0;            /* 1 => signal has happened, needs attention */
int sigchld_flag = 0;           /* 1 => signal has happened, needs attention */
int sigalrm_flag = 0;           /* 1 => signal has happened, needs attention */
int sigterm_flag = 0;           /* lame duck mode */
time_t current_time;
int max_fd = 0;
int pending_requests = 0;

/* static to boa.c */
static void fixup_server_root(void);
static int create_server_socket(void);
static void drop_privs(void);

static int sock_opt = 1;
static int do_fork = 1;
int devnullfd = -1;

int main(int argc, char **argv)
{
    int c;                      /* command line arg */
    int server_s;                   /* boa socket */

    /* set umask to u+rw, u-x, go-rwx */
    c = umask(~0600);
    if (c == -1) {
        perror("umask");
        exit(1);
    }

    devnullfd = open("/dev/null", 0);

    /* make STDIN and STDOUT point to /dev/null */
    if (devnullfd == -1) {
        DIE("can't open /dev/null");
    }

    if (dup2(devnullfd, STDIN_FILENO) == -1) {
        DIE("can't dup2 /dev/null to STDIN_FILENO");
    }

    if (dup2(devnullfd, STDOUT_FILENO) == -1) {
        DIE("can't dup2 /dev/null to STDOUT_FILENO");
    }

    /* but first, update timestamp, because log_error_time uses it */
    (void) time(&current_time);

    while ((c = getopt(argc, argv, "c:r:d")) != -1) {
        switch (c) {
        case 'c':
            if (server_root)
                free(server_root);
            server_root = strdup(optarg);
            if (!server_root) {
                perror("strdup (for server_root)");
                exit(1);
            }
            break;
        case 'r':
            if (chdir(optarg) == -1) {
                log_error_time();
                perror("chdir (to chroot)");
                exit(1);
            }
            if (chroot(optarg) == -1) {
                log_error_time();
                perror("chroot");
                exit(1);
            }
            if (chdir("/") == -1) {
                log_error_time();
                perror("chdir (after chroot)");
                exit(1);
            }
            break;
        case 'd':
            do_fork = 0;
            break;
        default:
            fprintf(stderr, "Usage: %s [-c serverroot] [-r chroot] [-d]\n", argv[0]);
            exit(1);
        }
    }

    fixup_server_root();
    read_config_files();
    open_logs();
    server_s = create_server_socket();
    init_signals();
    drop_privs();
    create_common_env();
    build_needs_escape();

    if (max_connections < 1) {
        struct rlimit rl;

        /* has not been set explicitly */
        c = getrlimit(RLIMIT_NOFILE, &rl);
        if (c < 0) {
            perror("getrlimit");
            exit(1);
        }
        max_connections = rl.rlim_cur;
    }

    /* background ourself */
    if (do_fork) {
        switch(fork()) {
        case -1:
            /* error */
            perror("fork");
            exit(1);
            break;
        case 0:
            /* child, success */
            break;
        default:
            /* parent, success */
            exit(0);
            break;
        }
    }

    /* main loop */
    timestamp();

    status.requests = 0;
    status.errors = 0;

    start_time = current_time;
    select_loop(server_s);
    return 0;
}

static int create_server_socket(void)
{
    int server_s;

    server_s = socket(SERVER_AF, SOCK_STREAM, IPPROTO_TCP);
    if (server_s == -1) {
        DIE("unable to create socket");
    }

    /* server socket is nonblocking */
    if (set_nonblock_fd(server_s) == -1) {
        DIE("fcntl: unable to set server socket to nonblocking");
    }

    /* close server socket on exec so cgi's can't write to it */
    if (fcntl(server_s, F_SETFD, 1) == -1) {
        DIE("can't set close-on-exec on server socket!");
    }

    /* reuse socket addr */
    if ((setsockopt(server_s, SOL_SOCKET, SO_REUSEADDR, (void *) &sock_opt,
                    sizeof (sock_opt))) == -1) {
        DIE("setsockopt");
    }

    /* internet family-specific code encapsulated in bind_server()  */
    if (bind_server(server_s, server_ip) == -1) {
        DIE("unable to bind");
    }

    /* listen: large number just in case your kernel is nicely tweaked */
    if (listen(server_s, backlog) == -1) {
        DIE("unable to listen");
    }
    return server_s;
}

static void drop_privs(void)
{
    /* give away our privs if we can */
    if (getuid() == 0) {
        struct passwd *passwdbuf;
        passwdbuf = getpwuid(server_uid);
        if (passwdbuf == NULL) {
            DIE("getpwuid");
        }
        if (initgroups(passwdbuf->pw_name, passwdbuf->pw_gid) == -1) {
            DIE("initgroups");
        }
        if (setgid(server_gid) == -1) {
            DIE("setgid");
        }
        if (setuid(server_uid) == -1) {
            DIE("setuid");
        }
        /* test for failed-but-return-was-successful setuid
         * http://www.securityportal.com/list-archive/bugtraq/2000/Jun/0101.html
         */
        if (setuid(0) != -1) {
            DIE("icky Linux kernel bug!");
        }
    } else {
        if (server_gid || server_uid) {
            log_error_time();
            fprintf(stderr, "Warning: "
                    "Not running as root: no attempt to change"
                    " to uid %d gid %d\n", server_uid, server_gid);
        }
        server_gid = getgid();
        server_uid = getuid();
    }
}

/*
 * Name: fixup_server_root
 *
 * Description: Makes sure the server root is valid.
 *
 */

static void fixup_server_root()
{
    char *dirbuf;

    if (!server_root) {
#ifdef SERVER_ROOT
        server_root = strdup(SERVER_ROOT);
        if (!server_root) {
            perror("strdup (SERVER_ROOT)");
            exit(1);
        }
#else
        fputs("boa: don't know where server root is.  Please #define "
              "SERVER_ROOT in boa.h\n"
              "and recompile, or use the -c command line option to "
              "specify it.\n", stderr);
        exit(1);
#endif
    }

    if (chdir(server_root) == -1) {
        fprintf(stderr, "Could not chdir to \"%s\": aborting\n",
                server_root);
        exit(1);
    }

    dirbuf = normalize_path(server_root);
    free(server_root);
    server_root = dirbuf;
}

